Milestone 3: Game logic

LoboArkano opened the pull request:
Develop rules to establish win and draw conditions, prevent a player from choosing an invalid position. Also prints and updates the board after a player's choice.
Created Board class:
- Methods: example, instructions, update_move, row_position and column_position

Created Player class:
- Methods: N/A

Created Rules class:
- Methods: in_range?, untaken?, previous_choice?, valid_move, win_check and draw_check

LoboArkano committed:
Merge branch 'feature/user_interface' into develop

Do it.

LoboArkano committed:
Merge branch 'develop' of github.com:LoboArkano/Tic-Tac-Toe into develop

LoboArkano committed:
Create classes for game.rb

- Board class
- Player class
- Rules class

LoboArkano committed:
Display example board

LoboArkano committed:
Create clear.rb file

Module to clean the screen

LoboArkano committed:
Update class Player

- Add reader for name
- Add choices attribute

LoboArkano committed:
Update main.rb

Use class Player to display name's player

LoboArkano committed:
Add instruction board

LoboArkano committed:
Display game board

LoboArkano committed:
Save choice of player

LoboArkano committed:
Show update board

Create update_move method

LoboArkano committed:
Use attr_reader :board

LoboArkano committed:
Update update_move method

Create two methods

LoboArkano committed:
Update update_move method

LoboArkano committed:
Update mage.rb

- Add require 'set'
- Create vistory_rows variable
- Create untaken? method
- Create valid_move method
- Create win_check method

LoboArkano committed:
Change control flow structure

LoboArkano committed:
Create draw_check method

LoboArkano committed:
Fix bug

Fix bug with include? method

LoboArkano committed:
Initialize variables for a new match

LoboArkano committed:
Update main.rb

reorder conditionals

LoboArkano committed:
Update game.rb

LoboArkano committed:
Update main.rb

Use proc to print the board

LoboArkano committed:
Update main.rb

Delete two variables

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
+      print sub_arr
+    end
+  end
+end
```
This should not be here but on the board class. Please create a method there and you can call it there, after.

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
-or all nine squares are filled.', ''
+or all nine squares are filled.'
+game_board.example
+print_board.call(game_board.board)
```
Here, instead of calling the proc, you call the method.

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
 print 'Player 2 Name: '
 input = gets.chomp
-p_names[:p2] = input
+p2 = Player.new(input)
```
You are not validating the players' names. If I click `enter`, I can add two empty names to the game and that is not good in terms of the UX.

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
+    if game_rules.win_check(p1.choices)
+      victory = true
+      winner = p1.name
     end
```
This should be in a winner method on the player Class. You can leave the if, but put the rest of the code on a method, cause you are calling it twice, so there is no need to repeat it.

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
+    if game_rules.win_check(p2.choices)
+      victory = true
+      winner = p2.name
+    end
```
Call the winner method mentioned above.

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
+    p1.choices.push(input)
+    puts "#{p1.name} choose #{p1.choices[-1]}", ''
+    game_board.update_move(p1.choices[-1].to_i, 'X')
+    turn += 1
```
You are repeating this code for player 1 and player 2. It is better to create a method and then call it for both of the players to keep your code DRY.

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
+    p2.choices.push(input)
+    puts "#{p2.name} choose #{p2.choices[-1]}", ''
+    game_board.update_move(p2.choices[-1].to_i, 'O')
+    turn += 1
```
use the method commented on lines 56-60 instead of repeating your code

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
-    if winner == true
-      puts "#{p2_name} Win the Game!"
+  draw = game_rules.draw_check(turn)
+  print_board.call(game_board.board)
```
Call the method that prints the board instead of the proc.
The board you print when the game has started doesn't have the numbers on each cell. It's easier for the players if you give them the numbers, so they can choose the cell easily.

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
-# Display default board
+puts 'Choice a number between 1 and 9'
+game_board.instructions
+print_board.call(game_board.board)
```
Call the method that prints the board instead of the proc.

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
+  draw = game_rules.draw_check(turn)
+  print_board.call(game_board.board)
+  puts '', ''
+  if victory || draw # rubocop:disable Style/Next
```
please don't disable rubocop. Instead, try to understand what you are not doing in the best way and fix it. Remember that the purpose of the linters is to remember you the best practices.

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
+    if victory == true
+      puts "#{winner} Wins the Game!"
       puts 'Congratulation!', ''
     elsif draw == true
```
There is no need to check line 80 and then check on line 81 and 84. You know that your `victory` and `draw` variables exist and that they are false, so just check if they are `true` instead of creating a nested `if statement`.

To make your code better (and not have accessible variables like victory and draw), you can create a method to check and another to modify those variables and just return their value, so then you can check here.

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
     puts 'You wanna play again? [Y/N]', ''
     input = gets.chomp
-    if input.include?('N' || 'n')
+    if input.include?('N') || input.include?('n')
       game_on = false
-    elsif input.include?('Y' || 'y')
+    elsif input.include?('Y') || input.include?('y')
```
This should be another method. You can call it check_restart. You can put it on the game Class and call it here.
There is no need to check if it's an upper case or lower case if you convert the input to one of them from the beginning.

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
-      p2_choices = []
       turn = 1
-      winner, draw = false
+      victory, draw = false
+      p1.choices = []
+      p2.choices = []
+      game_board = Board.new
     else
       game_on = false
     end
```
This should be a private method on the Game Class that you call on check_restart. You can call it restart_game.

majovanilla commented on lib/game.rb:
```diff
@@ -0,0 +1,100 @@
+  end
+end
+
+class Player
+  attr_reader :name
+  attr_accessor :choices
+  def initialize(name)
+    @name = name
+    @choices = []
+  end
+end
```
Create a file for each class and put it in the lib folder. That keeps your code more organized.

majovanilla commented on lib/game.rb:
```diff
@@ -0,0 +1,100 @@
+  end
+end
+
+class Rules
+  def initialize
+    @victory_rows = [%w[1 2 3], %w[4 5 6], %w[7 8 9], %w[1 4 7], %w[2 5 8], %w[3 6 9], %w[1 5 9], %w[3 5 7]]
+  end
+
+  def in_range?(choice)
+    choice.to_i >= 1 && choice.to_i <= 9
+  end
+
+  def untaken?(choice, opponent_choices)
+    arr = []
+    arr.push(choice)
+    return false if arr.to_set.subset?(opponent_choices.to_set)
+
+    true
+  end
+
+  def previous_choice?(choice, choices)
+    !choices.include?(choice)
+  end
+
+  def valid_move(choice, choices, opponent_choices)
+    in_range?(choice) && untaken?(choice, opponent_choices) && previous_choice?(choice, choices)
+  end
+
+  def win_check(choices)
+    @victory_rows.each do |victory_row|
+      return true if victory_row.to_set.subset?(choices.to_set)
+    end
+    false
+  end
+
+  def draw_check(turn)
+    turn > 9
+  end
+end
```
Create a file for each class and put it in the lib folder. That keeps your code more organized.

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
 
 puts '', "Let's Start!", ''
-# Display default board
+puts 'Choice a number between 1 and 9'
```
It should say: 'Choose a number  between 1 and 9'

majovanilla commented on bin/main.rb:
```diff
@@ -1,74 +1,103 @@
     elsif draw == true
       puts "It's a draw!", ''
     end
     puts 'You wanna play again? [Y/N]', ''
     input = gets.chomp
```
When you choose to play again, the board is not displayed for the first move. Please make the adjustments to print it so the players know what to click.

majovanilla reviewed (changes requested):
Hey @LoboArkano , 👋

## You did a *great* job on this project! 💜 

There are still some changes 👇 you need to implement to approve this request 🔍

### Required Changes: ✍️

- You should keep your code more organized. I gave you some tips below to achieve this. Please implement them.

- Remember to keep most of your code inside methods, to avoid repetition and also for security reasons. It's always better to create one method and call it multiple times than put the same code multiple times. I gave you some instructions on this, too.

## Please, make all the changes requested on the comments below and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

LoboArkano committed:
Create file board.rb

LoboArkano committed:
Delete proc in main.rb

Now is a Board method

LoboArkano committed:
Validate names

Not accept empty lines
Not only space names
Delete spaces at the beginning and at the end

LoboArkano committed:
Fix bug with strip method

LoboArkano committed:
Simplify untaken? method

LoboArkano committed:
Create player.rb file

LoboArkano committed:
Create method to end the game

LoboArkano committed:
Create update_game method

LoboArkano committed:
Change control flow

Rename game_rules for game

LoboArkano committed:
Update main.rb

Use upcase for the final input

LoboArkano committed:
Improve UI

LoboArkano committed:
Delete unnecessary variable

LoboArkano committed:
Update main.rb

kobiyoyo commented on lib/game.rb:
```diff
@@ -0,0 +1,73 @@
+
+  def update_game(player, input, game_board, sign)
+    player.choices.push(input)
+    puts "#{player.name} choose #{player.choices[-1]}"
```
No puts/print and get.chomp outside main.rb file

kobiyoyo commented on lib/board.rb:
```diff
@@ -0,0 +1,61 @@
+
+  def show
+    @board.each do |arr|
+      puts
+      arr.each do |sub_arr|
+        print sub_arr
+      end
```
No puts/print and get.chomp outside main.rb file.

kobiyoyo reviewed (changes requested):
## Hi @LoboArkano      ![](https://user-images.githubusercontent.com/38801423/62908122-703d4480-bd44-11e9-9115-42025ab7f341.png) 
#### Great Job so far, Just few things to work on. 
### General
- [x] The last move isnt displayed on the board lets fix that.
![1](https://user-images.githubusercontent.com/23278161/80647445-63918580-8a66-11ea-8a08-d8d25508aa9c.png)
Check comments below
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

LoboArkano committed:
Show board with the last move

LoboArkano committed:
Create print_board method

LoboArkano committed:
Delete show method

LoboArkano committed:
Delete print/puts from update_game method

maelfosso commented on lib/game.rb:
```diff
@@ -0,0 +1,71 @@
+class Game
+  attr_accessor :victory, :turn, :end_game
+  def initialize
+    @victory_rows = [%w[1 2 3], %w[4 5 6], %w[7 8 9], %w[1 4 7], %w[2 5 8], %w[3 6 9], %w[1 5 9], %w[3 5 7]]
```
**`[OPTIONAL]`** It could be better to define it as a CONSTANT

maelfosso reviewed (approved):
# Congratulations :clap: 

It's Ok :+1: You can merge it now :smile: 
Good continuation :handshake: 

Mael FOSSO
